src/pymor/algorithms/ml/vkoga.py

```diff
@@ -0,0 +1,426 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+from scipy.linalg import solve_triangular
+
+from pymor.algorithms.greedy import WeakGreedySurrogate, weak_greedy
+from pymor.core.base import BasicObject
+from pymor.core.defaults import defaults
+
+
+class VKOGASurrogate(WeakGreedySurrogate):
+    """Surrogate for the weak greedy error used when training the :class:`VKOGAEstimator`.
+
+    Not intended to be used directly.
+    """
+
+    def __init__(self, kernel, X_train, F_train, criterion='fp', reg=1e-12):
+        assert criterion in ('fp', 'f', 'p')
+        self.__auto_init(locals())
+        self.X_train = np.asarray(X_train)
+        self.F_train = np.asarray(F_train)
+        if self.F_train.ndim == 1:
+            self.F_train = self.F_train.reshape((-1, 1))
+        self.N, self.m = self.F_train.shape
+
+        self._centers = None
+        self._centers_idx = None
+        self._coefficients = None
+        self._L = None  # block-Cholesky factor of the kernel matrix
+        self._init_power_function_evals()
+
+    def predict(self, X):
+        X = np.asarray(X)
+        if self._centers is None:
+            return np.zeros((X.shape[0], self.m))
+        y = np.zeros((X.shape[0], self.m))
+        for j, cj in enumerate(self._centers):
```

Comment:
I think this can be done without a for-loop, as the implementation of the kernel is vectorized correctly.